core/rs/core/src/tableinfo.rs

```diff
@@ -0,0 +1,261 @@
+         FROM pragma_table_info('{table}') ORDER BY cid ASC"
+    );
+    let column_infos = match db.prepare_v2(&sql) {
+        Ok(stmt) => {
+            let mut cols: Vec<crsql_ColumnInfo> = vec![];
+
+            while stmt.step()? == ResultCode::ROW {
+                cols.push(crsql_ColumnInfo {
+                    type_: stmt.column_text(2).map_err(free_cols(&cols))?.into_c_ptr(),
+                    name: stmt.column_text(1).map_err(free_cols(&cols))?.into_c_ptr(),
+                    notnull: stmt.column_int(3)?,
+                    cid: stmt.column_int(0)?,
+                    pk: stmt.column_int(4)?,
+                });
+            }
+
+            if cols.len() != columns_len {
+                err.set(&"Number of fetched columns did not match expected number of columns");
+                return Err(free_cols(&cols)(ResultCode::ERROR));
+            }
+            cols
+        }
+        Err(code) => {
+            err.set(&format!("Failed to prepare select for crr -- {table}"));
+            return Err(code);
+        }
+    };
+
+    let (mut pks, non_pks): (Vec<_>, Vec<_>) =
+        column_infos.clone().into_iter().partition(|x| x.pk > 0);
+    pks.sort_by_key(|x| x.pk);
+
+    unsafe {
+        *table_info = crsql_TableInfo {
+            baseCols: column_infos.into_c_ptr(),
+            baseColsLen: columns_len as i32,
+            tblName: table.into_c_ptr(),
+            nonPksLen: non_pks.len() as i32,
+            nonPks: non_pks.into_c_ptr(),
+            pksLen: pks.len() as i32,
+            pks: pks.into_c_ptr(),
+        }
+        .into_c_ptr();
+    }
+
+    return Ok(ResultCode::OK);
+}
+
+fn free_cols<'a>(cols: &'a Vec<crsql_ColumnInfo>) -> impl Fn(ResultCode) -> ResultCode + 'a {
+    move |err: ResultCode| {
+        for info in cols {
+            drop(unsafe { CString::from_raw(info.type_) });
+            drop(unsafe { CString::from_raw(info.name) });
+        }
+        err
+    }
+}
+
+pub unsafe fn free_table_info(table_info: *mut crsql_TableInfo) {
+    if table_info.is_null() {
+        return;
+    }
+
+    let info = *table_info;
+    if !info.tblName.is_null() {
+        drop(CString::from_raw(info.tblName));
+    }
+    if !info.baseCols.is_null() {
+        free_cols(&Vec::from_raw_parts(
+            info.baseCols,
+            info.baseColsLen as usize,
+            info.baseColsLen as usize,
+        ))(ResultCode::OK);
+    }
+    if !info.pks.is_null() {
+        drop(Vec::from_raw_parts(
+            info.pks,
+            info.pksLen as usize,
+            info.pksLen as usize,
+        ));
+    }
+    if !info.nonPks.is_null() {
+        drop(Vec::from_raw_parts(
+            info.nonPks,
+            info.nonPksLen as usize,
+            info.nonPksLen as usize,
+        ));
+    }
+    drop(Box::from_raw(table_info));
+}
+
+pub fn is_table_compatible(db: *mut sqlite::sqlite3, table: &str, err: *mut *mut c_char) -> bool {
```

Comment:
This should return Result<bool, ResultCode> given it can fail for reasons besides the table not being compatible.